parser_library/src/workspaces/workspace.cpp

```diff
@@ -475,66 +491,82 @@ std::regex pathmask_to_regex(std::string input)
-            add_diagnostic(diagnostic_s::error_L0001(root));
+            add_diagnostic(diagnostic_s::error_L0001(root.to_presentable()));
         return;
     }
 
+    std::regex path_validator = pathmask_to_regex(path_pattern.get_uri());
+
+    std::set<utils::resource::resource_location> processed_dirs;
+    std::deque<utils::resource::resource_location> dirs_to_search;
+
+    dirs_to_search.emplace_back(root);
+
     constexpr size_t limit = 1000;
     while (!dirs_to_search.empty())
     {
-        if (processed_canonical_paths.size() > limit)
+        if (processed_dirs.size() > limit)
         {
-            add_diagnostic(diagnostic_s::warning_L0005(path_pattern, limit));
+            add_diagnostic(diagnostic_s::warning_L0005(path_pattern.to_presentable(), limit));
             break;
         }
-        auto [path, canonical_path] = std::move(dirs_to_search.front());
-        dirs_to_search.pop_front();
 
-        if (!processed_canonical_paths.insert(std::move(canonical_path)).second)
+        const auto& dir = dirs_to_search.front();
+
+        if (!processed_dirs.insert(dir).second)
             continue;
 
-        if (std::regex_match(path, path_validator))
-            prc_grp.add_library(
-                std::make_unique<library_local>(file_manager_, utils::resource::resource_location(path), opts));
-
-        auto rc = utils::path::list_directory_subdirs_and_symlinks(
-            path, [&processed_canonical_paths, &dirs_to_search](const auto& path) {
-                std::error_code ec;
-                auto cp = utils::path::canonical(path, ec).string();
-                if (ec || processed_canonical_paths.count(cp))
-                    return;
-                if (utils::path::is_directory(cp))
-                    dirs_to_search.emplace_back(fix_path(path.string()), std::move(cp));
-            });
-        if (rc != utils::path::list_directory_rc::done)
+        if (std::regex_match(dir.get_uri(), path_validator))
         {
-            add_diagnostic(diagnostic_s::error_L0001(path));
+            prc_grp.add_library(std::make_unique<library_local>(file_manager_, dir, opts));
+        }
+
+        auto [subdir_list, return_code] = file_manager_.list_directory_subdirs_and_symlinks(dir);
+        if (return_code != utils::path::list_directory_rc::done)
+        {
+            add_diagnostic(diagnostic_s::error_L0001(dir.to_presentable()));
             break;
         }
+
+        for (auto& [_, subdir] : subdir_list)
+        {
+            if (processed_dirs.contains(subdir))
+                continue;
+
+            dirs_to_search.emplace_back(subdir);
+        }
+
+        dirs_to_search.pop_front();
     }
 }
 
 namespace {
+utils::resource::resource_location transform_to_resource_location(
+    std::string path, const utils::resource::resource_location& base_resource_location)
+{
+    utils::resource::resource_location rl;
+
+    if (std::filesystem::path fs_path = path; utils::path::is_absolute(fs_path))
```

Comment:
what if path is url? should that be tested first?

Reply:
I suggest leaving it in this order but adding the following chceck: `utils::path::is_directory(fs_path))`.

What do you think?

Reply:
I think when going with URL checking first, the path "C:/dir/file" would be considered a url when it shouldn't be, right? Or what am I missing?

Reply:
I see... however, there is one more case to consider - could the path constructor throw on a non-path argument?

Reply:
Well, according to docu this is up to implementation...

So, maybe, surround the `path` constructor with `try - catch` block. Inside the `try` block also check for `utils::path::is_directory(fs_path))`.

If the above doesn't throw and `is_directory` is `true` -> this is a filesystem path,
otherwise continue with the rest of the logic


---------------------------------------
This can potentially be enhanced by doing the above only for 1 letter schemes.

Reply:
I think that the easiest option is for the `is_uri` to simply say false on windows for c:/ and similar "urls".

Reply:
Agreed and modified accordingly.